lib/atscppapi/examples/websocket/WebSocket.cc

```diff
@@ -0,0 +1,246 @@
+                "Upgrade: websocket\r\n"
+                "Connection: Upgrade\r\n"
+                "Sec-WebSocket-Accept: " + digest + "\r\n\r\n";
+            produce(headers);
+            ws_handshake_done_ = true;
+        }
+    }
+
+    // auto tid = std::this_thread::get_id();
+    if (type == InterceptPlugin::REQUEST_HEADER) {
+        headers_ += data;
+    } else {
+        // cout << tid << ": Read request body" << endl << data << endl;
+        if (isWebsocket()) {
+
+	    // When we've read as much as we can for any incoming
+	    // data chunk we remove the head of ws_buf_ and reset pos_
+	    // to 0.
+	    //
+	    // If we can't read a complete length + masks, we leave
+	    // pos_ unchanged and just append what we got to ws_buf_.
+	    //
+	    // When we get the length, we update pos_ and length.
+	    // While length is non-zero we wait for ws_buf_ to have
+	    // msg_len_ bytes beyond pos_.
+
+            ws_buf_ += data;
+            while (true) {
+                if (!msg_len_) {
+                    size_t avail = ws_buf_.size() - pos_;
+
+                    // Read the msg_length if we have enough data.
+                    if (avail < 6) break; // 2 + 4 mask
+                    ctrl_ = ws_buf_[pos_] & 0xF;
+                    size_t msg_len = ws_buf_[pos_ + 1] & 0x7F;
+                    if (msg_len == 0x7E) {
+                        if (avail < 8) { // 2 + 2 + length bytes + 4 mask.
+                            break;
+                        }
+                        msg_len = 0;
+                        pos_ += 2;
+                        for (int i = 0; i < 2; ++i, ++pos_) {
+                            msg_len <<= 8;
+                            msg_len += (unsigned char)(ws_buf_[pos_]);
+                        }
+                    } else if (msg_len == 0x7F) {
+                        if (avail < 14) { // 2 + 8 length bytes + 4 mask.
+                            break;
+                        }
+                        msg_len = 0;
+                        pos_ += 2;
+                        for (int i = 0; i < 8; ++i, ++pos_) {
+                            msg_len <<= 8;
+                            msg_len += (unsigned char)(ws_buf_[pos_]);
+                        }
+                    } else {
+                        pos_ += 2;
+                    }
+                    msg_len_ = msg_len;
+                    // Copy the mask.
+                    for (size_t i = 0; i < 4; ++i, ++pos_) {
+                        mask_[i] = ws_buf_[pos_];
+                    }
+                }
+
+                // Check if we have enough data to read the message.
+                if (ws_buf_.size() < pos_ + msg_len_) break; // not enough data.
+
+                // Apply the mask.
+                for (size_t i=0; i<msg_len_; ++i) {
+                    ws_buf_[pos_ + i] ^= mask_[i & 3];
+                }
+
+                std::string message = ws_buf_.substr(pos_, msg_len_);
+
```

Comment:
For what it's worth you're not responding to PING frames (0x9), clients will likely close on you.